### PR TITLE
Parallel testing

### PR DESCRIPTION
**Pull request by eileencodes**
@tenderlove and I worked on this which adds parallel testing to Rails applications by default. New applications will have parallel testing enabled by default, and older applications can add it to their test helper:

```ruby
class ActiveSupport::TestCase
  parallelize(workers: 2)
end
```

Parallel testing in this implementation utilizes forking processes over threads. The reason we (tenderlove and eileencodes) chose forking processes over threads is forking will be faster with single databases, which most applications will use locally. Using threads is beneficial when tests are IO bound but the majority of tests are not IO bound. NOTE: after some experimentation we added a threaded parallelization method as well, but forked processes are still the default.

If an application doesn't want to use parallel testing they can either remove the `parallelize` block from the test application or set `PARALLEL_WORKERS` to `1` or fewer. For environments where you want to change the default number of workers from what you've set in your `test_helper.rb` you can export / set an environment variable to change the number of workers used. The following will use 15 workers and split the tests into 15 processes.

```
PARALLEL_WORKERS=15 bin/rails test
```

Note: ~~While parallel testing will work with multiple primary databases, it currently doesn't rollback fixtures correctly. I'm actively working on fixing that but decided it was out of scope for this particular feature, since fixing it is not a feature of parallel testing but rather a bug / inconsistency in how Rails is handled. The fix for that should be coming shortly. Parallel testing and multiple primary databases does work with tests if not using fixtures.~~ I'm not sure why I thought this but I just tested it locally again and the fixtures work. I think I had a bug in my setup the last time I tested this.

If you have multiple databases they can be setup like this in your `test_helper.rb`

```ruby
class ActiveSupport::TestCase
  parallelize(workers: 2)

  parallelize_setup do |worker|
    # create a db w/ worker. Runs after processes are forked
  end

  parallelize_teardown do |worker|
    # delete the test databases or other cleanup. Runs before processes are closed
  end
end
```

To do:
- [x] Documentation
- [x] Guides
- [x] CHANGELOG

cc/ @tenderlove @dhh

### CODE REVIEW

**Comment by tenderlove** [comment]
This looks great!  3 things off the top of my head:

1. I'm not sure if Windows supports unix sockets. We should test that and possibly opt-out Windows folks.
2. JRuby definitely doesn't support `fork`.  We might be able to get this to work with `Process.spawn` out of the box, but maybe we should opt-out JRuby folks for now too?
3. I like the class-level `parallelize` API for setting the number of workers, but do we care if people call the method more than once? (I think the answer is "no", but I just want to check)

**Comment by nynhex** [comment]
This is great work!  Thanks @eileencodes @tenderlove ❤️

**Comment by jasl** [comment]
It seems Windows 10 upcoming RS4 release will partially support unix sockets, see <https://blogs.msdn.microsoft.com/commandline/2017/12/19/af_unix-comes-to-windows/>

**Comment by metacritical** [comment]
Does anyone use rails on Windows either development or production? MRI is still large i dont know anyone using JRUby on production. Opting them out for this release is OK.

**Comment by GBH** [comment]
Is this a replacement for https://github.com/grosser/parallel_tests ?

**Comment by eileencodes** [comment]
@metacritical I don't want to discount users of JRuby or Windows, however I think since we're a ways out from Rails 6.0 there's a ton of potential for this feature to evolve. I'd love to see a way to use threads over processes if someone finds that useful. I don't think it's necessary for the first iteration though. Once we have an API nailed down and this merged we can iterate on it and add a threads feature.

@GBH yes technically it replaces it, but not because we think the gem is doing anything incorrect. We're not personally using it at GitHub but we wrote our own implementation based on how we parallelize our own tests.

File: activerecord/lib/active_record/test_databases.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    def self.create_and_migrate(i, spec_name:)
+      old, ENV["VERBOSE"] = ENV["VERBOSE"], "false"
+
+      connection_spec = ActiveRecord::Base.configurations[spec_name]
+
+      connection_spec["database"] += "-#{i}"
+      ActiveRecord::Tasks::DatabaseTasks.create(connection_spec)
+      ActiveRecord::Base.establish_connection(connection_spec)
+      if ActiveRecord::Base.connection.migration_context.needs_migration?
+        ActiveRecord::Tasks::DatabaseTasks.migrate
```
Comment (bogdanvlviv) [inline activerecord/lib/active_record/test_databases.rb]: I think we should always execute (and rely on) `ActiveRecord::Tasks::DatabaseTasks.migrate` to ensure full setup for test database since this creates tables like `schema_migrations`, `ar_internal_metadata` etc.

File: activerecord/lib/active_record/test_databases.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      end
+    ensure
+      ActiveRecord::Base.establish_connection(ActiveRecord::Base.configurations[Rails.env])
```
Comment (bogdanvlviv) [inline activerecord/lib/active_record/test_databases.rb]: It is just question: can't realize why do we need re-establish the connection?

File: activerecord/lib/active_record/test_databases.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+require "active_support/testing/parallelization"
+
+module ActiveRecord
+  module TestDatabases
```
Comment (bogdanvlviv) [inline activerecord/lib/active_record/test_databases.rb]: I think we should express this module as private api by `# :nodoc:`

**Comment by alissonbrunosa** [comment]
I'm afraid that problem is beyond the scope, but is there any way to define whether a test should run on threads or processes? For instance, if I had a test that is IO bound, it would be awesome to run on threads.
I don’t know, maybe this will solve the JRuby’s problem with `fork` as well.

**Comment by yeongsheng-tan** [comment]
Nice and sweet. 👏💖

File: activerecord/lib/active_record/test_databases.rb (hunk 2 above)
Comment (eileencodes) [reply activerecord/lib/active_record/test_databases.rb]: We need to re-establish the original connection to AR Base since we want the `test` connection, not the other db connection. Otherwise the tests will try to all run against AR Base with the other db connection, not the `test` connection.

File: activerecord/lib/active_record/test_databases.rb (hunk 1 above)
Comment (eileencodes) [reply activerecord/lib/active_record/test_databases.rb]: They'll always need a migration because they'll always be brand new. I think however I'm going to drop this implementation and load up the structure/schema instead because it will be faster.

File: activerecord/lib/active_record/test_databases.rb (hunk 3 above)
Comment (eileencodes) [reply activerecord/lib/active_record/test_databases.rb]: I haven't decided yet if I want to make this private or document it. I think it could be useful for setting up test dbs when you have multiple databases.

File: activesupport/lib/active_support/testing/parallelization.rb [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+          @queue = Queue.new
+        end
+
+        def record(reporter, result)
+          reporter.synchronize do
+            reporter.record(result)
+          end
+        end
+
+        def <<(o)
```
Comment (simi) [inline activesupport/lib/active_support/testing/parallelization.rb]: What not use delegator for `<<` and `pop` methods?

**Comment by pangolingo** [comment]
Sometimes I set breakpoints in tests with Pry or Byebug. I assume the might not work with multiple processes. Could we add a note to the docs mentioning that? (Sorry I haven't had a chance to test this for myself.)

**Comment by yuki24** [comment]
@pangolingo when debugging with pry or byebug you should probably run a single test rather than all of them. But at least byebug works well in a multi-threaded environment (pauses other threads when it hits `byebug` and resumes upon `continue`). I'm not sure how it behaves in a forked process.

File: activesupport/lib/active_support/test_case.rb [hunk 5]
```diff
@@ -39,6 +40,66 @@ def test_order=(new_order)
...
+      #     end
+      #   end
+      def parallelize_setup(&block)
+        ActiveSupport::Testing::Parallelization.after_fork_hook do |worker|
+          yield worker
+        end
+      end
+
+      # Cleanup required for parallel testing. This can be used to drop databases
+      # if you're app uses multiple write/read databases or other clean up before
```
Comment (tjschuck) [inline activesupport/lib/active_support/test_case.rb]: `you're` => `your` 😅

File: guides/source/testing.md [hunk 6]
```diff
@@ -462,6 +462,37 @@ Rails options:
...
+
+To enable parallelization add the following to your `test_helper.rb`:
+
+```
+class ActiveSupport::TestCase
+  parallelize(workers: 2)
+end
+```
+
+The number of workers passes is the number of times the process will be forked. You may want to
```
Comment (tjschuck) [inline guides/source/testing.md]: `passes` => `passed`

File: guides/source/testing.md [hunk 7]
```diff
@@ -462,6 +462,37 @@ Rails options:
...
+
+```
+PARALLEL_WORKERS=15 bin/rails test
+```
+
+When parallelizing tests, Active Record automatically handles creating and migrating a database for each
+process. The databases will be suffixed with the number corresponding to the worker. For example, if you
+have 2 workers the tests will create `test-database-0` and `test-database-1` respectively.
+
+If the number of workers passes is 1 or fewer the processes will not be forked and the tests will not
```
Comment (tjschuck) [inline guides/source/testing.md]: `passes` => `passed`

File: activerecord/lib/active_record/test_databases.rb (hunk 3 above)
Comment (eileencodes) [reply activerecord/lib/active_record/test_databases.rb]: Decided to no doc this for now. Easier to document later than undocument.

**Comment by eileencodes** [comment]
Decided to add a threaded parallelizer after all. JRuby apps will automatically be generated using the threaded one. If you want to use threads just add `with: :threads` as a keyword argument.

I've updated docs, guides, and added a changelog.

File: activerecord/lib/active_record/test_databases.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+    def self.create_and_migrate(i, spec_name:)
+      old, ENV["VERBOSE"] = ENV["VERBOSE"], "false"
+
+      connection_spec = ActiveRecord::Base.configurations[spec_name]
+
+      connection_spec["database"] += "-#{i}"
+      ActiveRecord::Tasks::DatabaseTasks.create(connection_spec)
+      ActiveRecord::Base.establish_connection(connection_spec)
+      ActiveRecord::Tasks::DatabaseTasks.migrate
```
Comment (eileencodes) [inline activerecord/lib/active_record/test_databases.rb]: I plan on replacing this with a structure load or a straight copy of the database later on but currently structure load doesn't work with multiple databases so sticking with migrate for now.

File: activerecord/lib/active_record/test_databases.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    ActiveSupport::Testing::Parallelization.run_cleanup_hook do |i|
+      drop(i, spec_name: Rails.env)
+    end
+
+    def self.create_and_migrate(i, spec_name:)
+      old, ENV["VERBOSE"] = ENV["VERBOSE"], "false"
+
+      connection_spec = ActiveRecord::Base.configurations[spec_name]
+
+      connection_spec["database"] += "-#{i}"
```
Comment (simi) [inline activerecord/lib/active_record/test_databases.rb]: What about underscore `"_#{i}"` to keep file based databases using underscores in file names?

Comment (eileencodes) [reply activerecord/lib/active_record/test_databases.rb]: They're temporary databases, unless it's going to break file based dbs I don't think underscore vs dash is a big deal.

**Comment by IgorDobryn** [comment]
Awesome!

**Comment by Fudoshiki** [comment]
@eileencodes 
In rails 5.2
```ruby
# frozen_string_literal: true

ENV['RAILS_ENV'] ||= 'test'
require_relative '../config/environment'
require 'rails/test_help'
require 'rubocop/rake_task'

RuboCop::RakeTask.new


class ActiveSupport::TestCase
  # Run tests in parallel with specified workers
  parallelize(workers: 2)

  # Add more helper methods to be used by all tests here...
end
```

rails t show failed **rubocop** tests

Now 
```ruby
# frozen_string_literal: true

ENV['RAILS_ENV'] ||= 'test'
require_relative '../config/environment'
require 'rails/test_help'
require 'rubocop/rake_task'

RuboCop::RakeTask.new


class ActiveSupport::TestCase
  # Run tests in parallel with specified workers
  parallelize(workers: 2)

  # Add more helper methods to be used by all tests here...
end
```
rails t ignoring `RuboCop::RakeTask.new`

How use that now?

**Comment by eileencodes** [comment]
@Fudoshiki can you open a new issue explaining the problem you're having?  From your comment I don't understand what the issue is. Thanks!

File: activesupport/lib/active_support/test_case.rb [hunk 10]
```diff
@@ -39,6 +40,91 @@ def test_order=(new_order)
...
+
+        return if workers <= 1
+
+        executor = case with
+                   when :processes
+                     Testing::Parallelization.new(workers)
+                   when :threads
+                     Minitest::Parallel::Executor.new(workers)
+                   else
+                     raise ArgumentError, "#{with} is not a supported parallelization exectutor."
```
Comment (dapicester) [inline activesupport/lib/active_support/test_case.rb]: I believe this `exectutor` is a typo.

Comment (bogdanvlviv) [reply activesupport/lib/active_support/test_case.rb]: It was fixed by a4e226fb2d4ae533b6ac668c38b6923b5eb5db0a

**Comment by johnvross** [comment]
Someone asked if anyone is using windows for rails development. Yes. I work for a state agency and use the windows subsystem for linux on windows for all development since the fall creators update and they fixed filewatchers. Just FYI we are out here.

**Comment by chrishough** [comment]
Does this apply to both rspec and minitest?

**Comment by eileencodes** [comment]
Hey @johnvross I know you're out there and I'm sorry someone asked that question as it's not an opinion the core team holds. I think you should be supported through the threaded parallelizer, but I'm  not sure the Unix sockets that dRB relies on will work for you.

@chrishough this is using Minitest's parallel executor so I don't think it will work for rspec per say but we wrote the API in such a way that it's easy for us to add support for another parallelizer. This feature is still very new and it will be awhile before Rails 6 is released.

**Comment by chrishough** [comment]
Thanks @eileencodes.  I was hoping it would replace https://github.com/grosser/parallel_tests, and I am definitely curious to see how this plays out.

**Comment by WaKeMaTTa** [comment]
@eileencodes if we use `rspec` or `cucumber` how we can use this feature?

**Comment by thepeoplesbourgeois** [comment]
@WaKeMaTTa I think that's the topic of the comments a few notes above yours; my guess would be that an adapter for rspec/cucumber would need to be written for this parallelizer

**Comment by WaKeMaTTa** [comment]
@thepeoplesbourgeois you are right. Thanks

**Comment by atef-ds** [comment]
@eileencodes i have an issue with this solution:
sometimes we need to populate our test db with fake data before running the test => with parallel test our new db test are empty and we lost our fake data...
To fix it, i did: 
1) my original db called "db_test"  (contains some fake data),
i add a new one calle db_test_template: createdb -T db_test db_test_template
2) in test_helper:
parallelize(workers: 4)
parallelize_setup do |worker|
    configuration = ActiveRecord::Base.connection_config
    #delete db
    ActiveRecord::Tasks::DatabaseTasks.drop configuration.stringify_keys
    ActiveRecord::Base.establish_connection(configuration)

    configuration[:template] = ENV["POSTGRES_TEMPLATE"]
   #create a new db 
    ActiveRecord::Tasks::DatabaseTasks.create(configuration.stringify_keys)
  # now i have a db with data that i populated before in the original db
end
To run test :
POSTGRES_TEMPLATE=db_test_template rails test -v
it works well for me
@eileencodes can we considerate somehow that we need to keep data that we populated before?
